### PR TITLE
Fix various issues in serializer and emitter around target body selection.

### LINKED ISSUE #1821: Invariant: serialized 12 of 15
Repro:

```js
(function() {
  function URI(other) {
    if (other) {
      throw new Error();
    }
    this.foo = other.noSuchMethod();
  }

  function fn(arg) {
    var first = new URI(arg);
    return function() {
      new URI(first);
    };
  }

  __optimize(fn);

  module.exports = fn;
})();
```

Log:

```
=== visited but not serialized values
AbstractValue (no intrinsic name) (cannot print value)
  referenced by 1 scopes
      {"start":{"line":11,"column":11},"end":{"line":13,"column":5},"source":"/Users/gaearon/p/prepack/fb-www/input.js","filename":"/Users/gaearon/p/prepack/fb-www/input.js"}
AbstractValue (no intrinsic name) (cannot print value)
  referenced by 1 scopes
      {"start":{"line":11,"column":11},"end":{"line":13,"column":5},"source":"/Users/gaearon/p/prepack/fb-www/input.js","filename":"/Users/gaearon/p/prepack/fb-www/input.js"}
EmptyValue (no intrinsic name) undefined
  referenced by 1 scopes
      {"start":{"line":11,"column":11},"end":{"line":13,"column":5},"source":"/Users/gaearon/p/prepack/fb-www/input.js","filename":"/Users/gaearon/p/prepack/fb-www/input.js"}
```

Invariant:

https://github.com/facebook/prepack/blob/7b66b4386405d8f6efe29e8a8bf41d27b6dd6f2c/src/serializer/ResidualHeapSerializer.js#L2093

```
Invariant Violation: serialized 12 of 15
This is likely a bug in Prepack, not your code. Feel free to open an issue on GitHub.
    at invariant (/Users/gaearon/p/prepack/src/invariant.js:16:15)
    at ResidualHeapSerializer.serialize (/Users/gaearon/p/prepack/src/serializer/ResidualHeapSerializer.js:2093:17)
    at statistics.referenceCounts.measure (/Users/gaearon/p/prepack/src/serializer/serializer.js:226:15)
    at PerformanceTracker.measure (/Users/gaearon/p/prepack/src/statistics.js:100:14)
    at ast (/Users/gaearon/p/prepack/src/serializer/serializer.js:206:38)
    at statistics.total.measure (/Users/gaearon/p/prepack/src/serializer/serializer.js:138:17)
    at PerformanceTracker.measure (/Users/gaearon/p/prepack/src/statistics.js:100:14)
    at Serializer.init (/Users/gaearon/p/prepack/src/serializer/serializer.js:103:35)
    at prepackSources (/Users/gaearon/p/prepack/src/prepack-standalone.js:66:33)
    at compileSource (/Users/gaearon/p/prepack/scripts/debug-fb-www.js:91:18)
```

### PR DESCRIPTION
**Pull request by NTillmann**
Release notes: None

This fixes #1821.

Improve tracking of target bodies in Emitter,
which is necessary now that generators can have effects
and one should not attempt to emit into a body
with the wrong effects in place; removed some dubious code there.

It turned out that the exact placement of `Object.freeze` and similar statements
was always a bit fragile; this issue came out after fixing the Emitter, and so
this is made robust by keeping track of the
emission of all other object property mutations, and only emitting the
`Object.freeze` after they are all done.

When visiting modified property bindings, re-visit object in case it was
previously visited from a different scope.

For objects visited (only) in scopes where additional generator effects have been
applied, record additional information for serialization to ensure that these
objects get serialized in their creation scopes, so that the initial values
of the object properties are not polluted by the additional generator effects
of some nested scope.

Adding regression test.

### CODE REVIEW

**Comment by NTillmann** [comment]
This currently fails `fb-www 12` react tests because of output mismatch:
```
Expected value to equal:
      <div className="Rooty McRootface"><div>Hello  </div><div>World</div>!</div>
    Received:
      <div><div>Hello  </div><div>World</div>!</div>
```
It is not clear to me if that's really a bug; if it's a bug, any help in producing a smaller repro would be appreciated...

File: src/serializer/ResidualHeapSerializer.js [hunk 1]
```diff
@@ -2100,7 +2226,7 @@ export class ResidualHeapSerializer {
       !Array.from(this.serializedValues).every(val => this.residualValues.has(val))
     ) {
       this._logSerializedResidualMismatches();
-      invariant(false, "serialized " + this.serializedValues.size + " of " + this.residualValues.size);
+      //invariant(false, "serialized " + this.serializedValues.size + " of " + this.residualValues.size);
```
Comment (hermanventer) [inline src/serializer/ResidualHeapSerializer.js]: This is probably not intentional.

**Comment by gaearon** [comment]
@NTillmann I'll look at the test failure now

**Comment by gaearon** [comment]
@NTillmann 

It does seem to introduce a React regression.
Here's a minimal `input.js` for `yarn debug-fb-www`:

```js
require("react");

__evaluatePureFunction(function() {
  var React = require("react");

  function nullthrows(x) {
    if (x != null) {
      return x;
    }
    throw new Error('no');
  };

  function App(props) {
    nullthrows(props.className);
    return <div className={props.className} />;
  }

  __optimizeReactComponentTree(App, {
    firstRenderOnly: true
  });

  module.exports = App;
});
```

The output is (with my inline comments that show what's wrong):

```js
(function () {
  var $$0 = {
    enumerable: false,
    configurable: true,
    writable: true
  };

  var _$2 = this;

  var _$3 = _$2.Error;
  var _$4 = _$3.prototype;
  var _$5 = _$2.Object;
  var _$6 = _$5.defineProperty;

  var _0 = function (props, context) {
    var _4 = <div />;  // <------------ the return value is hoisted up and misses className

    var _6 = {
      className: void 0
    };

    var _8 = (__constructor.prototype = _$4, new __constructor());

    $$0.value = "Error\n    at nullthrows (/Users/gaearon/p/prepack/fb-www/input.js:10:11)\n    at App (/Users/gaearon/p/prepack/fb-www/input.js:14:5)", _$6(_8, "stack", $$0);
    $$0.value = "no", _$6(_8, "message", $$0);
    var _D = props;
    var _$0 = _D.className;

    var _1 = _$0 != null;

    if (_1) {
      var _$1 = _D.className;
      return _4;  // <--------------- if we didn't throw, we return _4
    } else {
      throw _8;
    }

    _4 = <div className={_$1} />; // <--------------- but it is set to the right thing too late!
    _6.className = _$1;
  };

  var __constructor = function () {};

  module.exports = _0;
}).call(this);
```

**Comment by NTillmann** [comment]
Awesome analysis, thanks @gaearon !

**Comment by NTillmann** [comment]
I've seen in other cases, including the new regression test added in this pull request, that all  modifications to properties are made explicit before a return-generator-entry via `ModifiedPropertyEntry`. I wonder why that didn't happen here.

**Comment by NTillmann** [comment]
Two more thoughts:
1) At best, this used to work only by chance.
2) Why is `_4` getting assigned twice? Sure, the second time looks like it was placed wrong. But no id should ever get assigned twice! Is there something react specific going on?

**Comment by gaearon** [comment]
Maybe JSX hoisting logic is buggy? Hoisting *is* React-specific.

**Comment by gaearon** [comment]
To be clear, it's not getting assigned twice on master:

```js
(function () {
  var $$0 = {
    enumerable: false,
    configurable: true,
    writable: true
  };

  var _$2 = this;

  var _$3 = _$2.Error;
  var _$4 = _$3.prototype;
  var _$5 = _$2.Object;
  var _$6 = _$5.defineProperty;

  var _0 = function (props, context) {
    var _D = props;
    var _$0 = _D.className;

    var _1 = _$0 != null;

    if (_1) {
      var _$1 = _D.className;

      var _4 = <div className={_$1} />;

      return _4;
    } else {
      var _8 = (__constructor.prototype = _$4, new __constructor());

      $$0.value = "Error\n    at nullthrows (/Users/gaearon/p/prepack/fb-www/input.js:10:11)\n    at App (/Users/gaearon/p/prepack/fb-www/input.js:14:5)", _$6(_8, "stack", $$0);
      $$0.value = "no", _$6(_8, "message", $$0);
      throw _8;
    }
  };

  var __constructor = function () {};

  module.exports = _0;
}).call(this);
```

**Comment by NTillmann** [comment]
That assignment looks indeed very suspicious and would explain the double-assignment. I guess @trueadm wrote that, maybe he can explain the intention and why it shouldn't lead to a double-assignment?

Maybe we can find a quick fix for this.

But I think we reached the point where @hermanventer and I should take a break from hacking, go to the whiteboard, and rethink what's happening with effectful generators in optimized functions.

**Comment by trueadm** [comment]
I'll add a way of disabling hoisting and look into this bug in ReactElement serializer.

**Comment by trueadm** [comment]
So after looking into it, the issue appears to be around the added:

`this.residualHeapSerializer.emitter.getBody()` to `ResidualReactElementSerializer`. It never used this before. The issue is that it's giving us the `AdditionalFunction` body, but this isn't the right body. The target body should be of that of the conditional. Currently it emits this:

```js
if (_1) {
  var _$1 = _S.className;
  return _4;
} else {
  throw _L;
}
_4 = <div className={_$1}>{_9}{_G}!</div>;
_6.className = _$1;
```

**Comment by trueadm** [comment]
I'll create a PR that splits the ReactElement visitor logic out and should make these problems easier to find.

**Comment by trueadm** [comment]
I've merge master into this branch and fixed merge conflicts.

**Comment by trueadm** [comment]
I've got a repro that isn't React specific, but shows the exact same problem:

```js
function nullthrows(x) {
  if (x != null) {
    return x;
  }
  throw new Error('no');
};

function App(props) {
  nullthrows(props.className);
  return {
    className: props.className,
  };
}

__optimize(App);
```

@NTillmann does that help any?

**Comment by trueadm** [comment]
I think I know what is going on now, let me try and patch your PR with a fix.

**Comment by trueadm** [comment]
The issue here seems to be this:

https://github.com/facebook/prepack/blob/ac4b8c6b0149a5b5780abc38f42e289d03ce4ef1/src/utils/generator.js#L361

When processing an `IfStatement`, we need to change the emitter target body to be that of the correct `consequentBody` or `alternateBody`. In the case above, when debugging, you can see it's not being changed (like it was before, by chance however) and as such the changes to the object get emitted to the wrong target body later on when we do `emitter._processCurrentBody`.

**Comment by trueadm** [comment]
@ntillmann the serializer output is broken. It’s assigning the className property after the return statement. This is just using the default test setup.

**Comment by NTillmann** [comment]
Got it. Yes, reproduce nicely. My `yarn watch` had stopped working...

**Comment by trueadm** [comment]
@hermanventer @NTillmann I stumbled across something. If I change `generator.purgeEntriesWithGeneratorDepencies` to be:

```js
  purgeEntriesWithGeneratorDepencies(): void {
    let newEntries = [];
    for (let oldEntry of this._entries) {
      if (oldEntry instanceof PossiblyNormalReturnEntry || oldEntry instanceof JoinedAbruptCompletionsEntry) continue;
      newEntries.push(oldEntry);
    }
    this._entries = newEntries;
  }
```

It seems both the above tests now pass.

**Comment by gaearon** [comment]
Accept to unblocked. Haven’t studied in detail but if the tests as passing (and the new nullthrows case is also fixed) this sounds great! I’ll read more tomorrow.

**Comment by trueadm** [comment]
I merged master and all tests are looking good. I want to merge this now as quite a few small bugs depend on the fixes in this PR.

**Comment by NTillmann** [comment]
I am trying to land...

**Comment by trueadm** [comment]
Opps, sorry, we did it exactly at the same time! :D

**Pull request closed without merging**